src/aspire/noise/noise.py

```diff
@@ -345,3 +354,297 @@ def estimate_noise_psd(self):
+        The samples to use in each image are given by `samples_idx` mask.
+        The correlation is computed up to a maximal distance of `max_d`.
+
+        Port of MATLAB `cryo_epsdR`.
+
+        :param images: `Image` instance
+        :param samples_idx: Boolean mask shaped `(L,L)`.
+        :param max_d: Max computed correlation distance in pixels.
+           Default of `None` yields `np.floor(L / 3)`.
+        :param batch_size:  The size of the batches in which to compute the variance estimate.
+        :return:
+            - Radial PSD array of shape
+            - Distances map array
+            - Count of nonzero correlations array
+        """
+
+        n_img = images.n_images
+        L = samples_idx.shape[-1]
+        batch_size = min(batch_size, n_img)
+
+        # Correlations more than `max_d` pixels apart are not computed.
+        if max_d is None:
+            max_d = np.floor(L / 3)
+        if max_d > L - 1:
+            logger.info(
+                f"`max_d` value {max_d}greater than number of image pixels {L}, clipping to {L-1}."
+            )
+        max_d = int(min(max_d, L - 1))
+
+        # Compute distances
+        # Note grid_2d['r'] is not used because we want an integer grid directly;
+        #   yields integer dists (radius**2) values.
+        X, Y = xp.mgrid[0 : max_d + 1, 0 : max_d + 1]
+        dists = X * X + Y * Y
+        dsquare = xp.sort(xp.unique(dists[dists <= max_d**2]))
+        x = xp.sqrt(dsquare)  # actual distances
+
+        # corrs[i] is the sum of all x[j]x[j+d] where d = x[i]
+        corrs = xp.zeros_like(dsquare, dtype=np.float64)
+        # corrcount[i] is the number of pairs summed in corr[i]
+        corrcount = xp.zeros_like(dsquare, dtype=np.int64)
+
+        # distmap maps [i,j] to k where dsquare[k] = i**2 + j**2.
+        #   -1 indicates distance is larger than max_d
+        distmap = xp.full(shape=dists.shape, fill_value=-1)
+
+        # This differs from the MATLAB code, avoids `bisect`.
+        for i, d in enumerate(dsquare):
+            inds = dists == d  # locations having distance `d`
+            distmap[inds] = i  # assign index into dsquare `i`
+        # From here on, distmap will be accessed with flat indices
+        distmap = distmap.flatten()
+        validdists = xp.argwhere(distmap != -1)
+
+        # Compute Ncorr using a constant unit image.
+        mask = xp.zeros((L, L))
+        mask[samples_idx] = 1
+        tmp = xp.zeros((batch_size, 2 * L + 1, 2 * L + 1))  # pad
+        tmp[0, :L, :L] = mask
+        # MATLAB code internally detects/implicitly casts,
+        #   we explicitly call rfft2/irfft2.
+        ftmp = fft.rfft2(tmp[0])
+        Ncorr = fft.irfft2(ftmp * ftmp.conj(), s=tmp.shape[1:])
+        Ncorr = Ncorr[: max_d + 1, : max_d + 1]  # crop
+        Ncorr = xp.round(Ncorr)
+
+        # Values of isotropic autocorrelation function
+        # R[i] is value of ACF at distance x[i]
+        R = xp.zeros(len(corrs))
```

Comment:
This is overwritten later.

Reply:
Yes, I changed that line at the end and forgot the rm, thanks!